_drafts/the-case-of-nservicebus-long-running-handlers-ocr-processing.md

```diff
@@ -0,0 +1,66 @@
+---
+typora-root-url: ..
+typora-copy-images-to: ../img/posts
+
+layout: post
+title: "The case of NServiceBus long running jobs: OCR Processing."
+author: Mauro Servienti
+synopsis: "Designing systems using a message based architecture is awesome. Messages are a nice way to design human interactions and to model how different components in a domain interact with each other. Unfortunately technology, sometimes, causes more headaches than needed. And when it comes to messaging, long running jobs are a interesting headache to deal with."
+tags:
+- NServiceBus
+- long running handlers
+- sagas
+---
+
+Designing systems using a message based architecture is awesome. Messages are a nice way to design human interactions and to model how different components in a domain interact with each other.
+Unfortunately technology, sometimes, causes more headaches than needed. And when it comes to messaging, long running jobs are a interesting headache to deal with.
+
+### OCR Processing
+
+Users need to process images and extract text using an optical character recognition system (OCR). This is done via a web interface that allows users to upload images for processing.
+
+Images processing takes time and is done in background, when processing is completed users are notified that their background jobs are ready to be consumed.
+
+This process can be outlined as follows:
+
+![1482055831647](/img\posts\1482055831647.png)
+
+OCR processing can take a long time and we don't want to hold the incoming user request till is completed. An interesting option is then to offload the OCR work to a back-end system:
+
+1. User request is received by the web application
+2. Web application sends a message on a queue to a back-end processing system
+3. Back-end processing system processes the image
+4. When processing is done an event (again a message) is sent thought the queuing system
+5. Web app reacts and notifies the user, e.g. via SignalR
+
+### Have we solved the issue?
+
+Not really. Back-end processing [3] cannot happen in the context of the incoming message, queuing systems have the concept of transactional processing that affects the time we have to process the incoming message.
+
+> There are transactional queuing systems that have to respect transaction timeouts and non-transactional queuing systems based on peek and lock (or similar) concepts meaning that a message, once picked by a processor is locked for a certain amount of time and then released to the queue if processing doesn't happen in that timeframe.
+
+In such a scenario increasing the transaction, or lock, timeout is not a solution, it's just a way to postpone the problem at a later time.
+
+### The state machine and the processor
+
+If we take a closer look at the business problem we are trying to solve we notice that we are mixing two different business concerns:
```

Comment:
instead of 
> If we take a closer look at the business problem we are trying to solve we notice that we are mixing two different business concerns:

try
> A closer look at the business problem shows that there are two different business concerns being mixed together